File: base92.py
# THE BEERWARE LICENSE (Revision 42):
# <thenoviceoof> wrote this file. As long as you retain this notice you
# can do whatever you want with this stuff. If we meet some day, and you
# think this stuff is worth it, you can buy me a beer in return
# - Nathan Hwang (thenoviceoof)

'''
base92: a library for encoding byte strings

>>> x = encode('hello world')
>>> x
'Fc_$aOTdKnsM#+'
>>> decode(x)
'hello world'

>>> y = encode('^\xb6;\xbb\xe0\x1e\xee\xd0\x93\xcb"\xbb\x8fZ\xcd\xc3')
>>> y
"C=i.w6'IvB/viUpRAw25"
>>> decode(y)
'^\\xb6;\\xbb\\xe0\\x1e\\xee\\xd0\\x93\\xcb"\\xbb\\x8fZ\\xcd\\xc3'
'''

import math

__version__ = (0, 0, 1)

def base92_chr(val):
    '''
    Map an integer value <91 to a char

    >>> base92_chr(0)
    '!'
    >>> base92_chr(1)
    '#'
    >>> base92_chr(61)
    '_'
    >>> base92_chr(62)
    'a'
    >>> base92_chr(90)
    '}'
    >>> base92_chr(91)
    Traceback (most recent call last):
        ...
    ValueError: val must be in [0, 91)
    '''
    if val < 0 or val >= 91:
        raise ValueError('val must be in [0, 91)')
    if val == 0:
        return '!'
    elif val <= 61:
        return chr(ord('#') + val - 1)
    else:
        return chr(ord('a') + val - 62)

def base92_ord(val):
    '''
    Map a char to an integer

    >>> base92_ord('!')
    0
    >>> base92_ord('#')
    1
    >>> base92_ord('_')
    61
    >>> base92_ord('a')
    62
    >>> base92_ord('}')
    90
    >>> base92_ord(' ')
    Traceback (most recent call last):
        ...
    ValueError: val is not a base92 character
    '''
    num = ord(val)
    if val == '!':
        return 0
    elif ord('#') <= num and num <= ord('_'):
        return num - ord('#') + 1
    elif ord('a') <= num and num <= ord('}'):
        return num - ord('a') + 62
    else:
        raise ValueError('val is not a base92 character')

def base92_encode(bytstr):
    '''
    Take a byte-string, and encode it in base 91

    >>> base92_encode("")
    '~'
    >>> base92_encode("\\x00")
    '!!'
    >>> base92_encode("\x01")
    '!B'
    >>> base92_encode("\xff")
    '|_'
    >>> base92_encode("aa")
    'D8-9~'
    >>> base92_encode("hello world")
    'Fc_$aOTdKnsM#+'
    >>> base92_encode("aaaaaaaaaaaaa")
    'D81RPya.)hgNA(%s'
    '''
    # always encode *something*, in case we need to avoid empty strings
    if not bytstr:
        return '~'
    # check if we need a lop char at the end
    lop = False
    if (8 * (len(bytstr) + 1) <
        13 * min(x for x in range(len(bytstr)+1) if 13*x >= 8*len(bytstr))):
        lop = True
    # prime the pump
    bitstr = '{:08b}'.format(ord(bytstr[0]))
    bytstr = bytstr[1:]
    resstr = ''
    while len(bytstr):
        while len(bitstr) < 13 and bytstr:
            bitstr += '{:08b}'.format(ord(bytstr[0]))
            bytstr = bytstr[1:]
        i = int(bitstr[:13], 2)
        resstr += base92_chr(i / 91)
        resstr += base92_chr(i % 91)
        bitstr = bitstr[13:]
    if bitstr:
        bitstr += '0' * (13 - len(bitstr))
        i = int(bitstr, 2)
        resstr += base92_chr(i / 91)
        resstr += base92_chr(i % 91)
    if lop:
        resstr += '~'
    return resstr

def base92_decode(bstr):
    '''
    Take a base92 encoded string, convert it back to a byte-string

    >>> base92_decode("")
    ''
    >>> base92_decode("~")
    ''
    >>> base92_decode("!!")
    '\\x00'
    >>> base92_decode("!B")
    '\\x01'
    >>> base92_decode("|_")
    '\\xff'
    >>> base92_decode("D8-9~")
    'aa'
    >>> base92_decode('Fc_$aOTdKnsM#+')
    'hello world'
    >>> base92_decode("D81RPya.)hgNA(%s")
    'aaaaaaaaaaaaa'
    '''
    bitstr = ''
    resstr = ''
    # we always have pairs of characters
    for i in range(len(bstr)/2):
        x = base92_ord(bstr[2*i])*91 + base92_ord(bstr[2*i+1])
        bitstr += "{:013b}".format(x)
        while 8 <= len(bitstr):
            resstr += chr(int(bitstr[0:8], 2))
            bitstr = bitstr[8:]
    # check if we have a lop char
    if bstr and bstr[-1] == '~':
        resstr = resstr[:-1]
    return resstr

encode = base92_encode
decode = base92_decode

if __name__ == "__main__":
    import doctest
    doctest.testmod()

File: c/src/base92.c
// THE BEERWARE LICENSE (Revision 42):
// <thenoviceoof> wrote this file. As long as you retain this notice you
// can do whatever you want with this stuff. If we meet some day, and you
// think this stuff is worth it, you can buy me a beer in return
// - Nathan Hwang (thenoviceoof)

#include <base92.h>

#include <stdio.h>

unsigned char ENCODE_MAPPING[256] = (unsigned char[]){
        33, 35, 36, 37, 38, 39, 40, 41, 42, 43,
        44, 45, 46, 47, 48, 49, 50, 51, 52, 53,
        54, 55, 56, 57, 58, 59, 60, 61, 62, 63,
        64, 65, 66, 67, 68, 69, 70, 71, 72, 73,
        74, 75, 76, 77, 78, 79, 80, 81, 82, 83,
        84, 85, 86, 87, 88, 89, 90, 91, 92, 93,
        94, 95, 97, 98, 99, 100, 101, 102, 103, 104,
        105, 106, 107, 108, 109, 110, 111, 112, 113, 114,
        115, 116, 117, 118, 119, 120, 121, 122, 123, 124,
        125, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0
};
unsigned char DECODE_MAPPING[256] = (unsigned char[]){
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
        255, 255, 255, 0, 255, 1, 2, 3, 4, 5,
        6, 7, 8, 9, 10, 11, 12, 13, 14, 15,
        16, 17, 18, 19, 20, 21, 22, 23, 24, 25,
        26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
        36, 37, 38, 39, 40, 41, 42, 43, 44, 45,
        46, 47, 48, 49, 50, 51, 52, 53, 54, 55,
        56, 57, 58, 59, 60, 61, 255, 62, 63, 64,
        65, 66, 67, 68, 69, 70, 71, 72, 73, 74,
        75, 76, 77, 78, 79, 80, 81, 82, 83, 84,
        85, 86, 87, 88, 89, 90, 255, 255, 255, 255,
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
        255, 255, 255, 255, 255, 255
};

unsigned char base92chr_encode(unsigned char byt) {
        return ENCODE_MAPPING[byt];
}

unsigned char base92chr_decode(unsigned char byt) {
        return DECODE_MAPPING[byt];
}

unsigned char* base92encode(unsigned char* str, int len) {
        unsigned int i, j;       // i for raw, j for encoded
        unsigned int size;       // for the malloc
        unsigned long workspace; // bits holding bin
        unsigned short wssize;   // number of good bits in workspace
        int tmp;
        unsigned char c;
        unsigned char *res;
        
        if (len == 0) {
                return "~";
        }
        // precalculate how much space we need to malloc
        size = (len * 8) % 13;
        if (size == 0) {
                size = 2 * ((len * 8) / 13);
        } else if (size < 7) {
                size = 2 * ((len * 8) / 13) + 1;
        } else {
                size = 2 * ((len * 8) / 13) + 2;
        }
        printf("size: %d\n", size);
        // do the malloc, add space for a null byte
        res = (char*)malloc(sizeof(char) * (size + 1));
        workspace = 0;
        wssize = 0;
        j = 0;
        for (i = 0; i < len; i++) {
                printf("i: %d, j: %d\n", i, j);
                workspace = workspace << 8 | str[i];
                wssize += 8;
                if (wssize >= 13) {
                        tmp = (workspace >> (wssize - 13)) & 8191;
                        c = base92chr_encode(tmp / 91);
                        if (c == 0) {
                                // do something, illegal character
                                free(res);
                                return NULL;
                        }
                        res[j++] = c;
                        c = base92chr_encode(tmp % 91);
                        if (c == 0) {
                                // do something, illegal character
                                free(res);
                                return NULL;
                        }
                        res[j++] = c;
                        wssize -= 13;
                }
        }
        // encode a last byte
        if (0 < wssize && wssize < 7) {
                tmp = (workspace << (6 - wssize)) & 63;  // pad the right side
                c = base92chr_encode(tmp);
                if (c == 0) {
                        // do something, illegal character
                        free(res);
                        return NULL;
                }
                res[j] = c;
        } else if (7 <= wssize) {
                tmp = (workspace << (13 - wssize)) & 8191; // pad the right side
                c = base92chr_encode(tmp / 91);
                if (c == 0) {
                        // do something, illegal character
                        free(res);
                        return NULL;
                }
                res[j++] = c;
                c = base92chr_encode(tmp % 91);
                if (c == 0) {
                        // do something, illegal character
                        free(res);
                        return NULL;
                }
                res[j] = c;
        }
        // add the null byte
        res[size] = 0;
        return res;
}

unsigned char* base92decode(unsigned char* str) {
        return "";
}
